sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/MetadataRequestRetryPolicy.java

```diff
@@ -0,0 +1,67 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.cosmos.implementation;
+
+import com.azure.cosmos.CosmosException;
+import com.azure.cosmos.implementation.directconnectivity.WebExceptionUtility;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import reactor.core.publisher.Mono;
+
+import java.net.URI;
+
+public class MetadataRequestRetryPolicy implements IRetryPolicy {
+
+    private final static Logger logger = LoggerFactory.getLogger(MetadataRequestRetryPolicy.class);
+    private final GlobalEndpointManager globalEndpointManager;
+    private RxDocumentServiceRequest request;
+
+    public MetadataRequestRetryPolicy(GlobalEndpointManager globalEndpointManager) {
+        this.globalEndpointManager = globalEndpointManager;
+    }
+
+    public void onBeforeSendRequest(RxDocumentServiceRequest request) {
+        this.request = request;
+    }
+
+    private static boolean shouldMarkRegionAsUnavailable(CosmosException exception) {
+
+        // check for network issues or connectivity issues
+        if (WebExceptionUtility.isNetworkFailure(exception)) {
+            return Exceptions.isSubStatusCode(exception, HttpConstants.SubStatusCodes.GATEWAY_ENDPOINT_UNAVAILABLE) ||
+                Exceptions.isSubStatusCode(exception, HttpConstants.SubStatusCodes.GATEWAY_ENDPOINT_READ_TIMEOUT);
+        }
+
+        return false;
+    }
+
+    @Override
+    public Mono<ShouldRetryResult> shouldRetry(Exception e) {
+
+        if (this.request == null) {
+            logger.error("onBeforeSendRequest has not been invoked with the MetadataRequestRetryPolicy...");
+            return Mono.just(ShouldRetryResult.error(e));
+        }
+
+        CosmosException cosmosException = Utils.as(e, CosmosException.class);
```

Comment:
do we need to validate whether exception is CosmosException?

Reply:
Fixed in next iteration.